multiclient/client.go

```diff
@@ -657,46 +657,69 @@ func (mc *Client) SubscribeNewHead(ctx context.Context, ch chan<- *Header) (ethe
 	}), nil
 }
 
-func (mc *Client) subscribeNewHead(ctx context.Context, wg *sync.WaitGroup, url string, ch chan<- *Header) error {
+func (mc *Client) subscribeNewHead(ctx context.Context, wg *sync.WaitGroup, id uint64, ch chan<- *Header) error {
+	logger := log.New("id", id)
 	defer wg.Done()
 
 	for {
-		rc := mc.rpcClientMap.Get(url)
-		if rc == nil {
-			log.Trace("EthClient has been removed", "url", url)
+		rc, existed := mc.rpcClientMap.GetById(id)
+		if !existed {
+			logger.Trace("EthClient has been removed")
 			return nil
 		}
-
-		headerCh := make(chan *types.Header)
-		c := ethclient.NewClient(rc)
-		sub, err := c.SubscribeNewHead(ctx, headerCh)
-		if err != nil {
-			log.Warn("Failed to subscribe new head", "url", url, "err", err)
+		if rc == nil {
+			logger.Warn("The client is not ready, wait for a while")
 		} else {
-		WAIT_NEW_HEADER:
-			for {
-				select {
-				case header := <-headerCh:
-					h := &Header{
-						Client: rc,
-						Header: header,
-					}
-					select {
-					case ch <- h:
-					case <-ctx.Done():
-						return nil
-					}
-				case err := <-sub.Err():
-					log.Warn("Failed during subscription", "url", url, "err", err)
-					break WAIT_NEW_HEADER
-				case <-ctx.Done():
-					return nil
-				}
+			// If we have error, we need to retry
+			err := doSubscribe(ctx, id, rc, ch)
+			if err == nil {
+				return nil
 			}
 		}
+
 		// retry subscribe after retryPeriod
-		time.Sleep(retryPeriod)
-		log.Trace("Retry to subscribe new head", "url", url)
+		t := time.NewTicker(retryPeriod)
+		select {
+		case <-t.C:
+		case <-ctx.Done():
+			t.Stop()
+			return nil
+		}
+		t.Stop()
+		logger.Trace("Retry to subscribe new head")
+	}
+}
+
+func doSubscribe(ctx context.Context, id uint64, rc *rpc.Client, ch chan<- *Header) error {
+	logger := log.New("id", id)
+	headerCh := make(chan *types.Header)
+	c := ethclient.NewClient(rc)
+	subCtx, cancel := context.WithCancel(ctx)
```

Comment:
Why we need create cancel context here?
SubscribeNewHead seems to be canceled only by client.

Reply:
Just ensure the `SubscribeNewHead` call is closed.